### PR TITLE
add findCache for memoryIdx

### PR DESCRIPTION
**Pull request by woodsaj**
Cache patterns and their results in a LRU.
When new defs are added to the index, only cached patterns that match
the new series name are invalidated.  The cache is purged after every
prune task runs and after any call to delete items from the index.

### CODE REVIEW

File: idx/memory/find_cache.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+func (c *FindCache) Get(orgId uint32, pattern string) ([]*Node, bool) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok {
+		findCacheMiss.Inc()
+		return nil, false
+	}
+	nodes, ok := cache.Get(pattern)
+	if !ok {
```
Comment (replay) [inline idx/memory/find_cache.go]: should this increase `findCacheMiss`?

Comment (woodsaj) [reply idx/memory/find_cache.go]: yes.

File: idx/memory/find_cache.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
```
Comment (robert-milan) [inline idx/memory/find_cache.go]: Is this also not a cache miss?

File: idx/memory/find_cache.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	c.RUnlock()
+	if !ok || cache.Len() < 1 {
+		return
+	}
+
+	select {
+	case c.newSeries[orgId] <- struct{}{}:
+	default:
+		c.Lock()
+		c.backoff[orgId] = time.Now().Add(findCacheBackoff)
```
Comment (replay) [inline idx/memory/find_cache.go]: is the primary reason for this backoff mechanism because we don't want to slow down the startup procedure when MT is adding a lot of new series? or is it expected that this backoff might occasionally also happen during run time? if it happens during run time we might want to get alerted on that, because it will probably cause the find performance to drop, so maybe we should have a metric that records when we decide to backoff...

Comment (robert-milan) [reply idx/memory/find_cache.go]: This only affects the index/find right? So, during startup shouldn't it just respond with a service unavailable until it is in a ready state?

I agree if it happens during run time it would probably be useful to get alerted.

Comment (replay) [reply idx/memory/find_cache.go]: If we only want to ensure that the startup doesn't get slowed down, then wouldn't it make more sense to just have an `Activate()` method on the cache which MT can call once the startup is complete? And then we don't need the whole backoff mechanism.
I'm a little worried about a scenario where all of a sudden the find latency jumps and it'll take some grepping through the logs to figure out why.

Comment (replay) [reply idx/memory/find_cache.go]: @robert-milan  `InvalidateFor()` gets called when metrics get added to the index, so on startup it would be called a lot of times, for no purpose because the cache will be empty: https://github.com/grafana/metrictank/pull/1233/files#diff-051eec914f9c2892a0d9f0d2a4d87336R465

Comment (woodsaj) [reply idx/memory/find_cache.go]: The purpose is to accommodate large bursts of new series.  If the findCache is full (default 1000 items), then every new series seen is compared against all 1000 cache patterns.  That is an expensive overhead for ingestion.  So, the moment a large burst of new series is detected (via not being able to process them fast enough), we just disable the cache for a while.

The backoff isn't really needed for startup, as the findCaches aren't initialized until queries are received.  MT instances dont accept queries until they have completed processing their backlog.

The backoff is more designed for customers sending lots of new series, eg. due to customers sending traffic for new datacenters or servers.

Comment (woodsaj) [reply idx/memory/find_cache.go]: The cacheHit ration will go to 0 when the cache is disabled, so we dont need any additional metric to track it.   An Info message is logged everytime that the cache is disabled.  So if the cacheHit ratio drops to 0, it would be easy to look at the logs and see why.

**Comment by replay** [comment]
Over the past few weeks we've been trying to make MT use less memory, while sacrificing a bit more CPU for that (f.e. object interning). This change is going into the opposite direction again (more mem / less cpu), so maybe we should disable the find cache by default and only enable it when we know we need it?

File: idx/memory/find_cache.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	return fc
+}
+
+func (c *FindCache) Get(orgId uint32, pattern string) ([]*Node, bool) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok {
+		findCacheMiss.Inc()
+		return nil, false
```
Comment (robert-milan) [inline idx/memory/find_cache.go]:
```suggestion
		return nil, ok
```

File: idx/memory/memory.go [hunk 5]
```diff
@@ -1411,8 +1441,8 @@ ORGS:
 			m.Unlock()
 			tl.Add(time.Since(lockStart))
 			pruned = append(pruned, defs...)
-
 		}
+		m.findCache.Purge(org)
```
Comment (replay) [inline idx/memory/memory.go]: similar like in the case of `.Delete()` this could be

```suggestion
		if len(pruned) {
		    m.findCache.Purge(org)
		}
```

File: idx/memory/find_cache.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	nodes, ok := cache.Get(pattern)
```
Comment (robert-milan) [inline idx/memory/find_cache.go]: Is there a chance that `cache` could have already been deleted before attempting this call in some fringe situations?

File: idx/memory/memory.go (hunk 5 above)
Comment (woodsaj) [reply idx/memory/memory.go]: that is already the case.  The loop exists early if `len(paths) == 0`

File: idx/memory/find_cache.go (hunk 6 above)
Comment (woodsaj) [reply idx/memory/find_cache.go]: It is possible that one thread calls `cache, ok := c.cache[orgId]` then releases the lock, then before `nodes, ok := cache.Get(pattern)` is executed another thread gets a write lock and calls `delete(c.cache, orgId)`.  But it doesnt really matter.  But this is still safe as the items in c.cache[orgId] are pointers.  The end result is that calls to `findCache.Get()` will return results based on the content of cache when the Rlock() was acquired an not on the contents of the cache at the specific time that `nodes, ok := cache.Get(pattern)` is executed.

**Comment by woodsaj** [comment]
> Over the past few weeks we've been trying to make MT use less memory, while sacrificing a bit more CPU for that (f.e. object interning). This change is going into the opposite direction again (more mem / less cpu), 

It is more memory, but not much, as the cache is just a map of strings, that point to slices of pointers.  Also, though heap memory could be potentially higher, instances with moderate query loads will see a reduction in allocations (which are needed to search the tree) which will help keep RSS lower.

File: idx/memory/find_cache.go (hunk 3 above)
Comment (replay) [reply idx/memory/find_cache.go]: The reason why `InvalidateFor()` is called on `.add()` is because the newly added metric might need to show up in the results of some common queries, right? So once the backoff time has passed and the cache gets reactivated again, isn't it going to contain stale entries that should have been invalidated during the time it was disabled.

**Comment by Dieterbe** [comment]
+1 to what Robert said.
any PR with an impact on performance should show relevant benchmark results.

**Comment by woodsaj** [comment]
Comparison between master and this branch for relevant Benchmarks
```
benchmark                                         old ns/op     new ns/op     delta
BenchmarkConcurrent4Find/partitioned-8            150287        40939         -72.76%
BenchmarkConcurrent4Find/unPartitioned-8          43931         22132         -49.62%
BenchmarkConcurrent8Find/partitioned-8            201209        37329         -81.45%
BenchmarkConcurrent8Find/unPartitioned-8          59867         19462         -67.49%
BenchmarkConcurrentInsertFind/partitioned-8       412081        42666         -89.65%
BenchmarkConcurrentInsertFind/unPartitioned-8     364459        20785         -94.30%

benchmark                                         old allocs     new allocs     delta
BenchmarkConcurrent4Find/partitioned-8            2324           257            -88.94%
BenchmarkConcurrent4Find/unPartitioned-8          637            187            -70.64%
BenchmarkConcurrent8Find/partitioned-8            2324           257            -88.94%
BenchmarkConcurrent8Find/unPartitioned-8          637            187            -70.64%
BenchmarkConcurrentInsertFind/partitioned-8       11788          258            -97.81%
BenchmarkConcurrentInsertFind/unPartitioned-8     10115          187            -98.15%

benchmark                                         old bytes     new bytes     delta
BenchmarkConcurrent4Find/partitioned-8            95997         27794         -71.05%
BenchmarkConcurrent4Find/unPartitioned-8          34411         18879         -45.14%
BenchmarkConcurrent8Find/partitioned-8            95987         27792         -71.05%
BenchmarkConcurrent8Find/unPartitioned-8          34411         18881         -45.13%
BenchmarkConcurrentInsertFind/partitioned-8       524699        27841         -94.69%
BenchmarkConcurrentInsertFind/unPartitioned-8     484934        18897         -96.10%
```

**Comment by Dieterbe** [comment]
this needs more comments about the design and implementation.
the backoff/InvalidateQueue stuff in particular is hard to follow.

File: idx/memory/memory.go [hunk 7]
```diff
@@ -453,6 +460,11 @@ func (m *UnpartitionedMemoryIdx) add(def *schema.MetricDefinition) idx.Archive {
 		return *archive
 	}
 
+	// invalidate any findCache items that match this series name
+	defer func() {
```
Comment (Dieterbe) [inline idx/memory/memory.go]: it should never be up to a call site to document what the function being called does.
so, remove this comment, and to `FindCache.InvalidateFor` add a comment explaining what exactly it does (which can be in more detail, seems like it clears every single subtree as well)

File: idx/memory/memory.go [hunk 8]
```diff
@@ -1192,7 +1220,9 @@ func (m *UnpartitionedMemoryIdx) Delete(orgId uint32, pattern string) ([]idx.Arc
 
 	statMetricsActive.Set(len(m.defById))
 	statDeleteDuration.Value(time.Since(pre))
-
+	if len(deletedDefs) > 0 {
+		m.findCache.Purge(orgId)
```
Comment (Dieterbe) [inline idx/memory/memory.go]: if `len(deletedDefs)` is small, it may make more sense to do a sequence of `InvalidateFor` calls instead of Purge ?

**Comment by Dieterbe** [comment]
Can we add a  fast-path for the find-cache disabled case? so it immediately returns out of Add/Purge/.. calls when disabled?

File: idx/memory/find_cache.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+			return
+		}
+		cache, err = lru.New(c.size)
+		if err != nil {
+			log.Errorf("memory-idx: findCache failed to create lru. err=%s", err)
+			return
+		}
+		c.Lock()
+		c.cache[orgId] = cache
+		c.newSeries[orgId] = make(chan struct{}, findCacheInvalidateQueue)
```
Comment (Dieterbe) [inline idx/memory/find_cache.go]: these newSeries chans smell a lot like util.Limiter, maybe we should use that to be more explicit ?

**Comment by woodsaj** [comment]
> Can we add a fast-path for the find-cache disabled case? so it immediately returns out of Add/Purge/.. calls when disabled?

we already do that.  Caches are per orgId, of `c.cache[orgId]` doesnt exist we immediately return.

File: idx/memory/find_cache.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+			"": {
+				Path:     "",
+				Children: make([]string, 0),
+				Defs:     make([]schema.MKey, 0),
+			},
+		},
+	}
+	pos := strings.Index(path, ".")
+	prevPos := 0
+	for {
```
Comment (Dieterbe) [inline idx/memory/find_cache.go]: all of the below logic could use some explaining. it's not immediately clear why an InvalidateFor would populate a tree structure

File: idx/memory/find_cache.go (hunk 9 above)
Comment (woodsaj) [reply idx/memory/find_cache.go]: util.Limiter is a blocking limiter.  We need a non-blocking one.

File: idx/memory/memory.go (hunk 8 above)
Comment (woodsaj) [reply idx/memory/memory.go]: updated

File: idx/memory/find_cache.go (hunk 10 above)
Comment (woodsaj) [reply idx/memory/find_cache.go]: updated

File: idx/memory/memory.go (hunk 7 above)
Comment (woodsaj) [reply idx/memory/memory.go]: updated

**Comment by Dieterbe** [comment]
> > Can we add a fast-path for the find-cache disabled case? so it immediately returns out of Add/Purge/.. calls when disabled?
>
> we already do that. Caches are per orgId, of c.cache[orgId] doesnt exist we immediately return.

this doesn't seem to be true. Add() will simply add the new cache it doesn't exist yet, so any subsequent calls will see a cache and operate on it.

**Comment by woodsaj** [comment]
> this doesn't seem to be true. Add() will simply add the new cache if it doesn't exist yet, so any subsequent calls will see a cache and operate on it.

https://github.com/grafana/metrictank/blob/master/idx/memory/find_cache.go#L76-L79

**Comment by Dieterbe** [comment]
Not sure what you're trying to say here. `t` would be `time.Time{}` there (== unix epoch) so that branch wouldn't be taken. what am i missing?

**Comment by woodsaj** [comment]
> `t` would be `time.Time{}`

If the cache is disabled, `t` will be a time in the future. https://github.com/grafana/metrictank/blob/master/idx/memory/find_cache.go#L138

**Comment by Dieterbe** [comment]
future proofing the above URL's:
https://github.com/grafana/metrictank/blob/707e6ce7df0c7cd53f73b75a282212f023ae9af8/idx/memory/find_cache.go#L76-L79
https://github.com/grafana/metrictank/blob/707e6ce7df0c7cd53f73b75a282212f023ae9af8/idx/memory/find_cache.go#L138

**Comment by Dieterbe** [comment]
we cleared it up in a call: the cache cannot be disabled.
